control-plane-operator/controllers/hostedcontrolplane/etcd/reconcile.go

```diff
@@ -1,209 +1,316 @@
+--initial-cluster-token=etcd-cluster \
+--initial-cluster=${INITIAL_CLUSTER} \
+--initial-cluster-state=new \
+--quota-backend-bytes=${QUOTA_BACKEND_BYTES} \
+--auto-compaction-mode=periodic \
+--auto-compaction-retention=0 \
+--peer-client-cert-auth=true \
+--peer-cert-file=/etc/etcd/tls/peer/peer.crt \
+--peer-key-file=/etc/etcd/tls/peer/peer.key \
+--peer-trusted-ca-file=/etc/etcd/tls/peer/peer-ca.crt \
+--client-cert-auth=true \
+--cert-file=/etc/etcd/tls/server/server.crt \
+--key-file=/etc/etcd/tls/server/server.key \
+--trusted-ca-file=/etc/etcd/tls/server/server-ca.crt
+`
 
-// etcdContainer is not a container that we build directly but is used
-// to assign scheduling/resources to it
-func etcdContainer() *corev1.Container {
-	return &corev1.Container{
-		Name: "etcd",
-	}
-}
+		var members []string
+		for i := 0; i < p.DeploymentConfig.Replicas; i++ {
+			name := fmt.Sprintf("etcd-%d", i)
+			members = append(members, fmt.Sprintf("%s=https://%s.etcd-discovery.%s.svc:2380", name, name, namespace))
+		}
+		initialCluster := strings.Join(members, ",")
 
-func buildEtcdOperatorContainer(image string) func(c *corev1.Container) {
-	return func(c *corev1.Container) {
-		c.Image = image
-		c.Command = []string{"etcd-operator"}
-		c.Args = []string{"-create-crd=false"}
+		c.Image = p.EtcdImage
 		c.ImagePullPolicy = corev1.PullAlways
+		c.Command = []string{"/bin/sh", "-c", script}
+		c.VolumeMounts = []corev1.VolumeMount{
+			{
+				Name:      "data",
+				MountPath: "/var/lib",
+			},
+			{
+				Name:      "peer-tls",
+				MountPath: "/etc/etcd/tls/peer",
+			},
+			{
+				Name:      "server-tls",
+				MountPath: "/etc/etcd/tls/server",
+			},
+			{
+				Name:      "client-tls",
+				MountPath: "/etc/etcd/tls/client",
+			},
+		}
 		c.Env = []corev1.EnvVar{
 			{
-				Name: "MY_POD_NAMESPACE",
+				Name: "POD_IP",
 				ValueFrom: &corev1.EnvVarSource{
 					FieldRef: &corev1.ObjectFieldSelector{
-						FieldPath: "metadata.namespace",
+						FieldPath: "status.podIP",
 					},
 				},
 			},
 			{
-				Name: "MY_POD_NAME",
+				Name: "NAMESPACE",
 				ValueFrom: &corev1.EnvVarSource{
 					FieldRef: &corev1.ObjectFieldSelector{
-						FieldPath: "metadata.name",
+						FieldPath: "metadata.namespace",
 					},
 				},
 			},
+			{
+				Name:  "INITIAL_CLUSTER",
+				Value: initialCluster,
+			},
+			{
+				Name:  "QUOTA_BACKEND_BYTES",
+				Value: strconv.FormatInt(p.StorageSpec.PersistentVolume.Size.Value(), 10),
+			},
+		}
+		c.Ports = []corev1.ContainerPort{
+			{
+				Name:          "client",
+				ContainerPort: 2379,
+				Protocol:      corev1.ProtocolTCP,
+			},
+			{
+				Name:          "peer",
+				ContainerPort: 2380,
+				Protocol:      corev1.ProtocolTCP,
+			},
+			{
+				Name:          "metrics",
+				ContainerPort: 2381,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		}
+		// TODO: Would it be correct for this to do a quorum check to induce rescheduling?
+		c.LivenessProbe = &corev1.Probe{
```

Comment:
Good points, and this has come up with @derekwaynecarr before... One question is what can we say about this piece:
>Pods get evacuated and rescheduled to a different node

I _thought_ Derek had some scenarios where eviction may not occur. Ring any bells? Any ways you can think of that removing the probe would break down here?

Reply:
And: for now, we know the times that etcd members have lost their identity: never, because we haven't implemented any code to make that happen (e.g. in response to data loss).

Reply:
I went ahead and removed both the liveness and startup probes, and we can revisit the topic later as we develop tests where they might play a role.